# Data exploration and assessment

## Airport data

In [4]:
import pandas as pd

airports_file = 'airport-codes_csv.csv'
airports_df = pd.read_csv(airports_file)

print(airports_df.shape)

(55075, 12)


In [7]:
# pandas interprets continent = NA (North America) as nulls
# we manually add these back in for this stage, but note that it will not be necessary in the ETL pipeline as PySpark 
# does not make the same mistake
airports_df.loc[airports_df['continent'].isna(), 'continent'] = 'NA'
airports_df.isna().sum()/airports_df.shape[0]

ident           0.000000
type            0.000000
name            0.000000
elevation_ft    0.127208
continent       0.000000
iso_country     0.004485
iso_region      0.000000
municipality    0.103059
gps_code        0.255016
iata_code       0.833155
local_code      0.479147
coordinates     0.000000
dtype: float64

#### Around 83% of entries have no iata_code value. This will be an intesive task to update without providing much value. We will drop this column. In addition, we will not need to include gps_code or local_code, so these will also be removed. 

In [8]:
# the iso_region column contains the US state code
# we will have to manuualy extract the two-letter code from the column
airports_df = airports_df.replace({'US-': ''}, regex=True)
airports_df.head()

ident           type                                name  elevation_ft  \
0   00A       heliport                   Total Rf Heliport          11.0   
1  00AA  small_airport                Aero B Ranch Airport        3435.0   
2  00AK  small_airport                        Lowell Field         450.0   
3  00AL  small_airport                        Epps Airpark         820.0   
4  00AR         closed  Newport Hospital & Clinic Heliport         237.0   

  continent iso_country iso_region  municipality gps_code iata_code  \
0        NA          US         PA      Bensalem      00A       NaN   
1        NA          US         KS         Leoti     00AA       NaN   
2        NA          US         AK  Anchor Point     00AK       NaN   
3        NA          US         AL       Harvest     00AL       NaN   
4        NA          US         AR       Newport      NaN       NaN   

  local_code                            coordinates  
0        00A     -74.93360137939453, 40.07080078125  
1       00AA                 -101.473911, 38.704022  
2       00AK            -151.695999146, 59.94919968  
3       00AL  -86.77030181884766, 34.86479949951172  
4        NaN                    -91.254898, 35.6087

## US cities demographics data

In [9]:
demographics_file = 'us-cities-demographics.csv'
demo_df = pd.read_csv(demographics_file, delimiter=';')

print(demo_df.shape)

(2891, 12)


#### There are only 16 rows with any nulls, covering 8 cities, all of which are for relatively small. We disclude these. 

In [13]:
demo_df[demo_df.isna().any(axis=1)]

City        State  Median Age  Male Population  \
111       San Juan  Puerto Rico        41.4         155408.0   
155         Caguas  Puerto Rico        40.4          34743.0   
258       Carolina  Puerto Rico        42.0          64758.0   
333   The Villages      Florida        70.5              NaN   
449   The Villages      Florida        70.5              NaN   
637       Carolina  Puerto Rico        42.0          64758.0   
1437  The Villages      Florida        70.5              NaN   
1747      San Juan  Puerto Rico        41.4         155408.0   
1748      Mayagüez  Puerto Rico        38.1          30799.0   
1995         Ponce  Puerto Rico        40.5          56968.0   
2004       Bayamón  Puerto Rico        39.4          80128.0   
2441      San Juan  Puerto Rico        41.4         155408.0   
2589      Guaynabo  Puerto Rico        42.2          33066.0   
2597        Caguas  Puerto Rico        40.4          34743.0   
2654      Guaynabo  Puerto Rico        42.2          33066.0   
2746      Mayagüez  Puerto Rico        38.1          30799.0   

      Female Population  Total Population  Number of Veterans  Foreign-born  \
111            186829.0            342237                 NaN           NaN   
155             42265.0             77008                 NaN           NaN   
258             77308.0            142066                 NaN           NaN   
333                 NaN             72590             15231.0        4034.0   
449                 NaN             72590             15231.0        4034.0   
637             77308.0            142066                 NaN           NaN   
1437                NaN             72590             15231.0        4034.0   
1747           186829.0            342237                 NaN           NaN   
1748            35782.0             66581                 NaN           NaN   
1995            64615.0            121583                 NaN           NaN   
2004            90131.0            170259                 NaN           NaN   
2441           186829.0            342237                 NaN           NaN   
2589            37426.0             70492                 NaN           NaN   
2597            42265.0             77008                 NaN           NaN   
2654            37426.0             70492                 NaN           NaN   
2746            35782.0             66581                 NaN           NaN   

      Average Household Size State Code                               Race  \
111                      NaN         PR                 Hispanic or Latino   
155                      NaN         PR                 Hispanic or Latino   
258                      NaN         PR  American Indian and Alaska Native   
333                      NaN         FL                 Hispanic or Latino   
449                      NaN         FL          Black or African-American   
637                      NaN         PR                 Hispanic or Latino   
1437                     NaN         FL                              White   
1747                     NaN         PR  American Indian and Alaska Native   
1748                     NaN         PR                              Asian   
1995                     NaN         PR                 Hispanic or Latino   
2004                     NaN         PR                 Hispanic or Latino   
2441                     NaN         PR                              Asian   
2589                     NaN         PR                 Hispanic or Latino   
2597                     NaN         PR  American Indian and Alaska Native   
2654                     NaN         PR  American Indian and Alaska Native   
2746                     NaN         PR                 Hispanic or Latino   

       Count  
111   335559  
155    76349  
258    12143  
333     1066  
449      331  
637   139967  
1437   72211  
1747    4031  
1748     235  
1995  120705  
2004  169155  
2441    2452  
2589   69936  
2597     624  
2654     589  
2746   65521

#### Each city has counts for between 1 and 5 races, which creates duplicate rows. For around 79% of the rows, all but two of the columns (Race and Count) are redundant. We'll create a new table which contains the race data, and remove it from the demographics table to reduce storage space required. 

In [14]:
min_races = min(demo_df.groupby(['City', 'State']).size())
max_races = max(demo_df.groupby(['City', 'State']).size())
print(min_races, max_races)

1 5


In [15]:
# Proportion of duplicated rows when Race and Count columns are removed
1-demo_df[['City', 'State']].drop_duplicates().shape[0]/demo_df.shape[0]

0.7938429609131789

## Temperature data

In [16]:
temperature_file = '../../data2/GlobalLandTemperaturesByCity.csv'
temp_df = pd.read_csv(temperature_file)

temp_df.head()

           dt  AverageTemperature  AverageTemperatureUncertainty   City  \
0  1743-11-01               6.068                          1.737  Århus   
1  1743-12-01                 NaN                            NaN  Århus   
2  1744-01-01                 NaN                            NaN  Århus   
3  1744-02-01                 NaN                            NaN  Århus   
4  1744-03-01                 NaN                            NaN  Århus   

   Country Latitude Longitude  
0  Denmark   57.05N    10.33E  
1  Denmark   57.05N    10.33E  
2  Denmark   57.05N    10.33E  
3  Denmark   57.05N    10.33E  
4  Denmark   57.05N    10.33E  
               dt  AverageTemperature  AverageTemperatureUncertainty     City  \
47555  1820-01-01               2.101                          3.217  Abilene   
47556  1820-02-01               6.926                          2.853  Abilene   
47557  1820-03-01              10.767                          2.395  Abilene   
47558  1820-04-01              17

In [17]:
print(min(temp_df['dt']), max(temp_df['dt']))

1743-11-01 2013-09-01


#### The temperature data we have covers 1743 to 2013, while the immigration dataset covers 2016 only. As there is no date overlap between these two, we cannot join on date. We instead take the latest 10 year's of temperature for each city going forward, averaged for each month. 

In [5]:
temp_df['dt'] = pd.to_datetime(temp_df['dt'], format='%Y/%m/%d')

In [6]:
temp_10_years_df = temp_df[temp_df['dt'] >= max(temp_df['dt']) - pd.offsets.DateOffset(years=10)]

In [7]:
temp_10_years_df['month'] = pd.DatetimeIndex(temp_10_years_df['dt']).month

temp_10_years_df.head()

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


dt  AverageTemperature  AverageTemperatureUncertainty   City  \
3118 2003-09-01              14.425                          0.172  Århus   
3119 2003-10-01               6.621                          0.467  Århus   
3120 2003-11-01               6.016                          0.295  Århus   
3121 2003-12-01               3.697                          0.284  Århus   
3122 2004-01-01              -0.660                          0.204  Århus   

      Country Latitude Longitude  month  
3118  Denmark   57.05N    10.33E      9  
3119  Denmark   57.05N    10.33E     10  
3120  Denmark   57.05N    10.33E     11  
3121  Denmark   57.05N    10.33E     12  
3122  Denmark   57.05N    10.33E      1

In [8]:
temp_10_years_df.groupby(['City','Country','Latitude','Longitude','month']).mean().reset_index()

City     Country Latitude Longitude  month  AverageTemperature  \
0      A Coruña       Spain   42.59N     8.73W      1              9.1581   
1      A Coruña       Spain   42.59N     8.73W      2              9.2842   
2      A Coruña       Spain   42.59N     8.73W      3             11.2129   
3      A Coruña       Spain   42.59N     8.73W      4             12.6073   
4      A Coruña       Spain   42.59N     8.73W      5             14.8401   
5      A Coruña       Spain   42.59N     8.73W      6             17.8176   
6      A Coruña       Spain   42.59N     8.73W      7             19.1891   
7      A Coruña       Spain   42.59N     8.73W      8             19.5752   
8      A Coruña       Spain   42.59N     8.73W      9             18.4079   
9      A Coruña       Spain   42.59N     8.73W     10             15.2289   
10     A Coruña       Spain   42.59N     8.73W     11             11.5217   
11     A Coruña       Spain   42.59N     8.73W     12              9.2684   
12       Aachen     Germany   50.63N     6.34E      1              1.8348   
13       Aachen     Germany   50.63N     6.34E      2              2.0777   
14       Aachen     Germany   50.63N     6.34E      3              5.3417   
15       Aachen     Germany   50.63N     6.34E      4              9.9743   
16       Aachen     Germany   50.63N     6.34E      5             13.4501   
17       Aachen     Germany   50.63N     6.34E      6             16.4441   
18       Aachen     Germany   50.63N     6.34E      7             18.5348   
19       Aachen     Germany   50.63N     6.34E      8             17.5637   
20       Aachen     Germany   50.63N     6.34E      9             14.7704   
21       Aachen     Germany   50.63N     6.34E     10             10.1066   
22       Aachen     Germany   50.63N     6.34E     11              6.2512   
23       Aachen     Germany   50.63N     6.34E     12              2.1398   
24      Aalborg     Denmark   57.05N    10.33E      1              0.9761   
25      Aalborg     Denmark   57.05N    10.33E      2              0.6895   
26      Aalborg     Denmark   57.05N    10.33E      3              2.9387   
27      Aalborg     Denmark   57.05N    10.33E      4              7.7232   
28      Aalborg     Denmark   57.05N    10.33E      5             12.1786   
29      Aalborg     Denmark   57.05N    10.33E      6             15.3243   
...         ...         ...      ...       ...    ...                 ...   
42090     Çorum      Turkey   40.99N    34.08E      7             21.7960   
42091     Çorum      Turkey   40.99N    34.08E      8             22.2835   
42092     Çorum      Turkey   40.99N    34.08E      9             17.6284   
42093     Çorum      Turkey   40.99N    34.08E     10             12.1567   
42094     Çorum      Turkey   40.99N    34.08E     11              5.8121   
42095     Çorum      Turkey   40.99N    34.08E     12              1.7175   
42096   Öskemen  Kazakhstan   50.63N    82.39E      1            -18.0060   
42097   Öskemen  Kazakhstan   50.63N    82.39E      2            -15.5082   
42098   Öskemen  Kazakhstan   50.63N    82.39E      3             -5.7268   
42099   Öskemen  Kazakhstan   50.63N    82.39E      4              5.3268   
42100   Öskemen  Kazakhstan   50.63N    82.39E      5             12.1160   
42101   Öskemen  Kazakhstan   50.63N    82.39E      6             17.6042   
42102   Öskemen  Kazakhstan   50.63N    82.39E      7             19.0120   
42103   Öskemen  Kazakhstan   50.63N    82.39E      8             16.8578   
42104   Öskemen  Kazakhstan   50.63N    82.39E      9             11.5362   
42105   Öskemen  Kazakhstan   50.63N    82.39E     10              4.3680   
42106   Öskemen  Kazakhstan   50.63N    82.39E     11             -4.7764   
42107   Öskemen  Kazakhstan   50.63N    82.39E     12            -13.9961   
42108    Ürümqi       China   44.20N    87.20E      1            -15.7229   
42109    Ürümqi       China   44.20N    87.20E      2            -11.0418   
42110  

## Immigration data

In [17]:
fname = '../../data/18-83510-I94-Data-2016/i94_jan16_sub.sas7bdat'
df = pd.read_sas(fname, 'sas7bdat', encoding="ISO-8859-1")

df.head()

(2847924, 28) (3742, 28)


#### Some columns in the immigrant data table are mostly (>95%) nulls, including 'occup', 'entdepu', and 'insnum'. We will disclude these

In [3]:
df.isna().sum()/df.shape[0]*100

cicid        0.000000
i94yr        0.000000
i94mon       0.000000
i94cit       0.000000
i94res       0.000000
i94port      0.000000
arrdate      0.000000
i94mode      0.004941
i94addr      5.171279
depdate      4.414819
i94bir       0.028001
i94visa      0.000000
count        0.000000
dtadfile     0.000063
visapost    60.509611
occup       99.556836
entdepa      0.004878
entdepd      4.333889
entdepu     99.988122
matflag      4.333889
biryear      0.028001
dtaddto      0.018340
gender      11.824817
insnum      95.661360
airline      3.217317
admnum       0.000000
fltno        0.388841
visatype     0.000000
dtype: float64

In [5]:
df.drop(['occup', 'entdepu', 'insnum'], axis=1)

cicid   i94yr  i94mon  i94cit  i94res i94port  arrdate  i94mode  \
0              2.0  2016.0     3.0   213.0   213.0     XXX  20543.0      NaN   
1              3.0  2016.0     3.0   245.0   245.0     XXX  20524.0      NaN   
2              7.0  2016.0     3.0   260.0   260.0     SFR  20514.0      1.0   
3              9.0  2016.0     3.0   691.0   691.0     FTL  20533.0      1.0   
4             11.0  2016.0     3.0   258.0   258.0     XXX  20514.0      1.0   
5             12.0  2016.0     3.0   101.0   101.0     NYC  20514.0      1.0   
6             14.0  2016.0     3.0   101.0   101.0     MIA  20514.0      1.0   
7             15.0  2016.0     3.0   101.0   101.0     NYC  20514.0      1.0   
8             17.0  2016.0     3.0   101.0   101.0     NYC  20514.0      1.0   
9             18.0  2016.0     3.0   101.0   101.0     NYC  20514.0      1.0   
10            20.0  2016.0     3.0   101.0   101.0     NYC  20514.0      1.0   
11            21.0  2016.0     3.0   101.0   101.0     NYC  20514.0      1.0   
12            22.0  2016.0     3.0   101.0   101.0     LOS  20514.0      1.0   
13            23.0  2016.0     3.0   101.0   101.0     CHI  20514.0      1.0   
14            24.0  2016.0     3.0   101.0   101.0     CHI  20514.0      1.0   
15            25.0  2016.0     3.0   101.0   101.0     SEA  20514.0      1.0   
16            26.0  2016.0     3.0   101.0   101.0     SEA  20514.0      1.0   
17            28.0  2016.0     3.0   101.0   117.0     ATL  20514.0      1.0   
18            29.0  2016.0     3.0   101.0   117.0     ATL  20514.0      1.0   
19            30.0  2016.0     3.0   101.0   117.0     ATL  20514.0      1.0   
20            31.0  2016.0     3.0   101.0   117.0     ATL  20514.0      1.0   
21            45.0  2016.0     3.0   103.0   103.0     ATL  20514.0      1.0   
22            46.0  2016.0     3.0   103.0   103.0     ATL  20514.0      1.0   
23            47.0  2016.0     3.0   103.0   103.0     ATL  20514.0      1.0   
24            48.0  2016.0     3.0   103.0   103.0     ATL  20514.0      1.0   
25            49.0  2016.0     3.0   103.0   103.0     ATL  20514.0      1.0   
26            53.0  2016.0     3.0   103.0   103.0     ATL  20514.0      1.0   
27            54.0  2016.0     3.0   103.0   103.0     BOS  20514.0      1.0   
28            55.0  2016.0     3.0   103.0   103.0     BOS  20514.0      1.0   
29            56.0  2016.0     3.0   103.0   103.0     BAL  20514.0      1.0   
...            ...     ...     ...     ...     ...     ...      ...      ...   
3157042  4046777.0  2016.0     3.0   745.0   745.0     PBB  20533.0      3.0   
3157043  5317403.0  2016.0     3.0   745.0   745.0     PBB  20538.0      3.0   
3157044  5767159.0  2016.0     3.0   165.0   745.0     PBB  20541.0      3.0   
3157045  2439630.0  2016.0     3.0   745.0   745.0     PBB  20526.0      3.0   
3157046  2225863.0  2016.0     3.0   745.0   745.0     PEM  20525.0      3.0   
3157047   161637.0  2016.0     3.0   745.0   745.0     PHU  20514.0      3.0   
3157048   506872.0  2016.0     3.0   745.0   745.0     PHU  20516.0      3.0   
3157049   911993.0  2016.0     3.0   745.0   745.0     PHU  20518.0      3.0   
3157050  2632338.0  2016.0     3.0   745.0   745.0     PHU  20527.0      3.0   
3157051  5043064.0  2016.0     3.0   745.0   745.0     PHU  20537.0      3.0   
3157052  5043065.0  2016.0     3.0   745.0   745.0     PHU  20537.0      3.0   
3157053  5317417.0  2016.0     3.0   745.0   745.0     PHU  20538.0      3.0   
3157054  5317416.0  2016.0     3.0   745.0   745.0     PHU  20538.0      3.0   
3157055   911911.0  2016.0     3.0   745.0   745.0     PHU  20518.0      3.0   
3157056  1273176.0  2016.0     3.0   745.0   745.0     SDP  20520.0      3.0   
3157057  4758673.0  2016.0     3.0   745.0   745.0     SDP  20536.0      3.0   
3157058  3511068.0  2016.0     3.0   745.0   745.0     SUM  20531.0      3.0   
3157059  1101170.0  2016.0     3.0   745.0   745.0     SYS  20519.0      3.0   
3157060  1101

In [4]:
x_df = df[df['i94port'] == 'XXX']
x_df.shape

(1725, 28)